### PR TITLE
Deep Copy utilities

### PR DESCRIPTION
**Pull request by md5sha256**
Adds a util class for deep copying `Collection`, `Map` and arrays. I haven't tested the changes but this should be simple enough that a quick glance will spot errors. If that's insufficient I can write a simple unit test for this.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Supplier;
+
+/**
+ * Utilities for deep-copying collections, maps and arrays
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:12]: Can you please add an `@author` tag here?

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+ */
+public final class CopyUtils {
+    
+    private CopyUtils() {}
+
+    /**
+     * Perform a deep copy of all the elements from a given {@link Collection} to another.
+     * <p>
+     * If the source collection contains <code>null</code> elements, the cloning function should be
+     * handle null input. Additionally, the sink should also accept any values
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:22]: * should **be able to**

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            T cloned = cloningFunction.apply(original);
+            sink.add(cloned);
+        }
+    }
+
+    /**
+     * Perform a deep copy of all the elements from a given {@link Collection} to another.
+     * <p>
+     * If the source collection contains <code>null</code> elements, the cloning function should be
+     * handle null input. Additionally, the sink supplied by the sink supplier
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:43]: * should **be able to**

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        C sink = sinkSupplier.apply(source.size());
+        deepCopy(source, cloningFunction, sink);
+        return sink;
+    }
+
+    /**
+     * Perform a deep copy of all the elements from a given {@link Map} to another.
+     * <p>
+     * If the source map contains <code>null</code> values, the cloning function should be
+     * handle null input. Additionally, the sink should also accept any keys and values
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:65]: * should **be able to**

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            V cloned = cloningFunction.apply(original);
+            sink.put(entry.getKey(), cloned);
+        }
+    }
+
+    /**
+     * Perform a deep copy of all the elements from a given {@link Map} to another.
+     * <p>
+     * If the source map contains <code>null</code> values, the cloning function should be
+     * handle null input. Additionally, the sink should also accept any keys and values
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:88]: * should **be able to**

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        M sink = sinkSupplier.get();
+        deepCopy(source, cloningFunction, sink);
+        return sink;
+    }
+
+    /**
+     * Perform a deep-clone transformation on all values in a given {@link Map}
+     * <p>
+     * If the source map contains <code>null</code> values, the cloning function should be
+     * handle null input
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:110]: * should **be able to**

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+            V cloned = cloningFunction.apply(original);
+            entry.setValue(cloned);
+        }
+    }
+
+    /**
+     * Perform a deep copy of all the elements from a given array to another.
+     * <p>
+     * If the source array contains <code>null</code> elements, the cloning function should be
+     * handle null input.
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:130]: * should **be able to**

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        for (int i = 0; i < source.length; i++) {
+            sink[i] = cloningFunction.apply(source[i]);
+        }
+    }
+
+    /**
+     * Perform a deep copy of all the elements from a given array to another.
+     * <p>
+     * If the source array contains <code>null</code> elements, the cloning function should be
+     * handle null input.
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:151]: * should **be able to**

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+     *
+     * @param source          The source of the elements
+     * @param cloningFunction The function which clones the elements
+     * @param sinkSupplier    The supplier which consumes the length of the source array and supplies 
+     *                        the array to copy the cloned elements to
+     * @param <T>             The type of elements in the arrays
+     *
+     */
+    public static <T> T[] deepCopy(@NonNull T[] source, @NonNull Function<T, T> cloningFunction, @NonNull IntFunction<T[]> sinkSupplier) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:160]: Should also annotate the output as `@Nonnull`

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+     * If the source array contains <code>null</code> elements, the cloning function should be
+     * handle null input.
+     *
+     * @param source          The source of the elements
+     * @param cloningFunction The function which clones the elements
+     * @param sink            The array in which to copy to the cloned elements to
+     * @param <T>             The type of elements in the arrays
+     *
+     */
+    public static <T> void deepCopy(@NonNull T[] source, @NonNull Function<T, T> cloningFunction, @NonNull T[] sink) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:138]: Should also annotate the output as `@Nonnull`

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+     * If the source map contains <code>null</code> values, the cloning function should be
+     * handle null input
+     *
+     * @param source          The source of the elements
+     * @param cloningFunction The function which clones the elements
+     * @param <K>             The type of keys in the map
+     * @param <V>             The type of values in the map
+     *
+     */
+    public static <K, V> void deepCopy(@NonNull Map<K, V> source, @NonNull Function<V, V> cloningFunction) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:118]: Should also annotate the output as `@Nonnull`

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+     * @param source          The source of the elements
+     * @param cloningFunction The function which clones the elements
+     * @param sinkSupplier    The supplier which consumes the size of the source map and 
+     *                        supplies the map to copy cloned elements to
+     * @param <K>             The type of keys in the maps
+     * @param <V>             The type of values in the maps
+     * @param <M>             The type of the returned map
+     *
+     */
+    public static <K, V, M extends Map<K, V>> M deepCopy(@NonNull Map<K, V> source, @NonNull Function<V, V> cloningFunction, @NonNull Supplier<M> sinkSupplier) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:100]: Should also annotate the output as `@Nonnull`

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+     *
+     * @param source          The source of the elements
+     * @param cloningFunction The function which clones the elements
+     * @param sink            The map in which to copy to the cloned elements to
+     * @param <K>             The type of keys in the maps
+     * @param <V>             The type of values in the maps
+     * @param <M>             The type of the sink
+     *
+     */
+    public static <K, V, M extends Map<K, V>> void deepCopy(@NonNull Map<K, V> source, @NonNull Function<V, V> cloningFunction, @NonNull M sink) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:76]: Should also annotate the output as `@Nonnull`

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+     * @param source          The source of the elements
+     * @param cloningFunction The function which clones the elements
+     * @param sinkSupplier    The supplier which consumes the size of the source collection and 
+     *                        supplies the collection to copy cloned elements to
+     *                        
+     * @param <T>             The type of elements in the collections
+     * @param <C>             The type of the returned collection
+     *
+     */
+    public static <T, C extends Collection<T>> C deepCopy(@NonNull Collection<T> source, @NonNull Function<T, T> cloningFunction, @NonNull IntFunction<C> sinkSupplier) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:55]: Should also annotate the output as `@Nonnull`

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+     * contained by the source collection, null or not.
+     *
+     * @param source          The source of the elements
+     * @param cloningFunction The function which clones the elements
+     * @param sink            The collection in which to copy to the cloned elements to
+     * @param <T>             The type of elements in the collections
+     * @param <C>             The type of the sink
+     *
+     */
+    public static <T, C extends Collection<T>> void deepCopy(@NonNull Collection<T> source, @NonNull Function<T, T> cloningFunction, @NonNull C sink) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:32]: Should also annotate the output as `@Nonnull`

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Couple of copy/pasted typos.

File: src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java (hunk 15 above)
Comment (md5sha256) [reply src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java:32]: This is a void method

**Comment by md5sha256** [comment]
Should be fixed

**Comment by md5sha256** [comment]
Actually, are the generics on `#deepClone(Collection, Function, Collection)` necessary? Maybe I can just have them as `Collection` instead.

**Comment by TheBusyBiscuit** [comment]
> Actually, are the generics on `#deepCopy(Collection, Function, Collection)` necessary? Maybe I can just have them as `Collection` instead. Same goes for `#deepCopy(Map, Function, Map)`
> 
> EDIT: Removed the unnecessary generics

Yeah, that is much better. The base type is more than enough here

**Review by TheBusyBiscuit — APPROVED** [review]

**Pull request merged**